### PR TITLE
custom rodio source for audio

### PR DESCRIPTION
**Pull request by tarkah**
This is very much an early draft and in need of feedback.

I've made `AudioOuput` able to play anything `Decodable`. `Decodable` has been implemented on `AudioSource` and is the concrete type used for creating the `AudioOutput` in the `AudioPlugin`. `Decodable` outputs a `Decoder` which must implement `rodio::Source<Item = rodio::Sample> + Send + Sync` to be accepted by `AudioOutput`.

The downside right now is if someone created their own `Decodable`, they'd need to setup their own `AudioOutput` on that type and redo a bunch of the stuff the `AudioPlugin` sets up on behalf of the user. I've done this at the following test with my Atra3+ decoder and it works. https://github.com/tarkah/bevy-test-custom-audio. Also, users have to specify `AudioOutput<AudioSource>` out of the box, which is less than ideal, though we could type alias something I'm sure.

I couldn't get dynamic dispatching working in the way I was hoping since `rodio::Sample` in the associated type for `rodio::Source` is `Sized`. I have no idea if it's possible to have a single `AudioOutput` with a dynamic `rodio::Source` input without having to define separate instances over the concrete `rodio::Source` types. I was hoping to be able to do a `Box<dyn Decodable>` type setup but it just doesn't work. I'm also a little out of my depth here so maybe I am missing something, or maybe I am just tackling this API change completely wrong. Anyway, your feedback would be appreciated!

### CODE REVIEW

**Comment by tarkah** [comment]
Also, I understand if this is super low priority, so no rush on this at all!

**Comment by multun** [comment]
Can you rebase this on master? A formatting change got through, and the CI doesn't check for it at the commit you're at

**Comment by tarkah** [comment]
Rebased

**Comment by multun** [comment]
Awesome, thanks! (the failure isn't yours)

**Comment by tarkah** [comment]
No problem! I noticed your PR to fix the formatting issue.

File: crates/bevy_audio/src/audio_output.rs [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
 use bevy_asset::{Assets, Handle};
 use bevy_ecs::Res;
-use rodio::{Decoder, Device, Sink};
-use std::{collections::VecDeque, io::Cursor, sync::RwLock};
+use rodio::{Device, Sink};
+use std::{collections::VecDeque, sync::RwLock};
 
 /// Used to play audio on the current "audio device"
-pub struct AudioOutput {
+pub struct AudioOutput<P>
```
Comment (cart) [inline crates/bevy_audio/src/audio_output.rs:8]: Can we make this `AudioOutput<P=AudioSource>` to make this a non-breaking change / retain the ease-of-use we had before?

File: examples/audio/audio.rs [hunk 2]
```diff
@@ -8,7 +8,7 @@ fn main() {
         .run();
 }
 
-fn setup(asset_server: Res<AssetServer>, audio_output: Res<AudioOutput>) {
+fn setup(asset_server: Res<AssetServer>, audio_output: Res<AudioOutput<AudioSource>>) {
```
Comment (cart) [inline examples/audio/audio.rs:11]: After making the change above, lets revert this to `Res<AudioOutput>`.

**Comment by tarkah** [comment]
@cart Thanks for taking some time to look at this. I understand this is an edge case, as most people are going to be using MP3, WAV, Vorbis or Flac sources that are supported by the `rodio::Decoder`. However, if someone wanted to play some other codec, this entire audio module would not work for them. For example, needing to play `.at3` that my [atrac3-plus decoder](https://github.com/tarkah/atrac3p-decoder) handles if you wanted to make an open source port of a PS Vita game using this library.

Since the `rodio::Source` trait is what enables all the awesome stuff rodio can do with sound, it just seems like a good idea to be able to support playing any `rodio::Source` and not just what can be decoded by `rodio::Decoder`.

Anyway, I'm still not in love with where I've landed on this API and I think it's reasonable to hold off on this until we are 100% sure it's a reasonable addition to the audio module. In the meantime I'll go ahead and rebase / make the changes you suggested.

**Review by cart — APPROVED** [review]

**Comment by cart** [comment]
I think this is in a good spot to merge. Thanks for your work here!

**Comment by tarkah** [comment]
Great, thanks @cart!

**Pull request merged**